docs/_ext/pharmpy_snippet.py

```diff
@@ -0,0 +1,199 @@
+        # self.state.nested_parse(self.content, self.content_offset, node)
+
+class PharmpyExecute(RecursiveDirective):
+    required_arguments = 0
+    optional_arguments = 0
+    final_argument_whitespace = True
+    has_content = True
+
+    option_spec = {
+        "hide-code": directives.flag,
+        "hide-output": directives.flag,
+        "code-below": directives.flag,
+        "linenos": directives.flag,
+        "lineno-start": directives.nonnegative_int,
+        "emphasize-lines": directives.unchanged_required,
+        "raises": csv_option,
+        "stderr": directives.flag,
+    }
+
+    def run(self):
+        # if 'hide-code' in self.options:
+            # raise ValueError('pharmpy-execute does not allow to hide code, use jupyter-execute instead')
+
+        return self._nodes()
+
+    def _nodes(self):
+        lines = self._lines()
+        return self._convert_lines_to_nodes(lines)
+
+    def _lines(self):
+        return [
+            f'.. container:: pharmpy-snippet{"" if "hide-output" in self.options else " with-output"}',
+            '',
+            *_indent(3, self._input_output_lines())
+        ]
+
+    def _input_output_lines(self):
+        if 'hide-code' in self.options:
+            return self._output()
+
+        if 'code-below' in self.options:
+            return [
+                *self._output(),
+                '',
+                *self._input(),
+            ]
+
+        return [
+            *self._input(),
+            '',
+            *self._output(),
+        ]
+
+    def _input(self):
+        return [
+            '.. tabs::',
+            *_indent(3, [
+                '',
+                '.. code-tab:: py',
+                *_indent(3, self._code_option_lines()),
+                '',
+                *_indent(3, self.content),
+                '',
+                '.. code-tab:: r R',
+                *_indent(3, self._code_option_lines()),
+                '',
+                *_indent(3, transpile_py_to_r(self.content)),
+            ]),
+        ]
+
+    def _code_option_lines(self):
+        if 'emphasize_lines' in self.options:
+            yield f':emphasize_lines:{self.options.get("emphasize_lines")}'
+        if "linenos" in self.options:
+            yield ':linenos:'
+        if "lineno-start" in self.options:
+            yield f':lineno-start:{self.options.get("lineno-start")}'
+
+    def _output(self):
+        return [
+            '.. jupyter-execute::',
+            *_indent(3, [
+                *self._jupyter_option_lines(),
+                '',
+                *self.content
+            ]),
+        ]
+
+    def _jupyter_option_lines(self):
+        yield ':hide-code:'
+        if "hide-output" in self.options:
+            yield ':hide-output:'
+        if 'raise' in self.options:
+            yield f':raises:{",".join(self.options.get("raises"))}'
+        if "stderr" in self.options:
+            yield ':stderr:'
+
+class PharmpyCode(RecursiveDirective):
```

Comment:
Maybe there would be a way to share more code between `PharmpyCode` and `PharmpyExecute`.

Reply:
Working on it.

Reply:
See https://github.com/pharmpy/pharmpy/pull/791/commits/08f529609f0ff4769e7c4a9426f1479e836b898c. What do you think?

Reply:
I think it looks good, but is there an advantage to have three classes (`PharmpyAbstractCodeDirective`, `PharmpyExecute`, `PharmpyCode`) instead of having `PharmpyCode` and `PharmpyExecute(PharmpyCode)`? Are there any instances of `PharmpyAbstractCodeDirective`?

Reply:
I think I just wanted to avoid overloading concrete implementations of methods. So I inherit from a common abstract ancestor.